src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2904,116 +2867,123 @@ private static bool IsReallyAType(TypeSymbol type)
+                    {
+                        continue;
+                    }
+                    TypeSymbolWithAnnotations source;
+                    TypeSymbolWithAnnotations destination;
+                    if (variance == VarianceKind.Out)
+                    {
+                        source = bound;
+                        destination = candidate;
+                    }
+                    else
+                    {
+                        source = candidate;
+                        destination = bound;
+                    }
+                    if (!ImplicitConversionExists(source, destination, ref useSiteDiagnostics, conversions.WithNullability(false)))
+                    {
+                        candidates.Remove(candidate);
+                    }
+                    else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames))
+                    {
+                        if (!TryMergeAndReplaceIfStillCandidate(candidates, candidate, bound, variance, conversions))
+                        {
+                            candidates.Remove(candidate);
+                        }
+                    }
                 }
-                list.Add(type);
             }
+        }
 
-            // Returns true if type was found and merged.
-            internal bool Merge(TypeSymbolWithAnnotations type, ConversionsBase conversions)
+        private static bool TryMergeAndReplaceIfStillCandidate(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            TypeSymbolWithAnnotations old,
+            TypeSymbolWithAnnotations @new,
+            VarianceKind variance,
+            ConversionsBase conversions)
+        {
+            // We make an exception when @new is dynamic, for backwards compatibility 
+            if (@new.IsDynamic())
             {
-                // We make an exception when type is dynamic, for backwards compatibility 
-                if (type.IsDynamic())
-                {
-                    return false;
-                }
-
-                List<TypeSymbolWithAnnotations> list;
-                if (!_dictionary.TryGetValue(type, out list))
-                {
-                    return false;
-                }
-
-                bool found = list.Contains(type, _valueComparer);
-                for (int i = 0; i < list.Count; i++)
-                {
-                    list[i] = MergeTupleNames(MergeDynamic(list[i], type, conversions.CorLibrary), type);
-                }
-                return found;
+                return true;
             }
 
-            internal void Remove(TypeSymbolWithAnnotations type)
+            if (candidates.TryGetValue(old, out TypeSymbolWithAnnotations latest))
             {
-                List<TypeSymbolWithAnnotations> list;
-                if (!_dictionary.TryGetValue(type, out list))
+                // PROTOTYPE(NullableReferenceTypes): `variance` seems questionable here. What if
```

Comment:
>PROTOTYPE(NullableReferenceTypes): [](start = 19, length = 34)

I assume this and other PROTOTYPE comments will be removed/converted as part of this PR.